packages/stack/src/views/Stack/CardContainer.tsx

```diff
@@ -152,8 +165,19 @@ function CardContainer({
     } else {
       onPageChangeCancel?.();
     }
-
-    onTransitionStart?.({ route }, closing);
+    /*fix for https://github.com/react-navigation/react-navigation/issues/9870
+       We trigger transition events only on the incoming (current) screen
+       */
+    if (previousRoute && closing) {
+      //route is closing send to the previous route for navigation.goBack() / navigation.pop / navigation.popToTop()
+      onTransitionStart?.({ route: previousRoute }, !closing);
+    } else {
+      //send to route for navigation.navigate  / navigation.push  / navigation.replace
+      // Run the operation in the next frame so that the target (new route) is available in useEventEmitter::emit.
+      requestAnimationFrame(() => {
+        onTransitionStart?.({ route }, closing);
+      });
```

Comment:
This is likely due to `componentDidMount` in `Card.tsx` firing before `useEffect` fires as it's async. So proper fix would be to refactor `Card.tsx` to function component so that we start animation in an `useEffect` to match it.

Reply:
There are multiple possible call stacks based on the kind of navigation action.

`Card.tsx`  `componentDidMount / componentDidUpdate / handleGestureStateChange`
`Card.tsx` `animate`
`CardContainer.tsx` `handleTransition`

`CardContainer.tsx` `handleTransition` is called directly from `Card.tsx`  `componentDidMount`  (via `Card.tsx` `animate`). Which `useEffect` are you referring to? the one in `CardContainer.tsx` isn't involved in the flow.

Card -> CardContainer callback registration uses a prop.

```
    <Card --snipped--
      onTransition={handleTransition}
```